### PR TITLE
LPS-131318 - Migrate frontend-js-tooltip-support-web to use typescript

### PR DESCRIPTION
**Pull request by bryceosterhaus**
Hey @wincent, I wanted to test out the typescript support and ported a module over following the docs you wrote up. Can you give me a sanity check on this just to be sure I'm not doing anything wrong? I'm not super confident on my ts here either, so if you notice anything let me know. I mostly just wanted to test this actually working and it seems to work when deployed for me.

### CODE REVIEW

File: modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx [hunk 1]
```diff
@@ -170,7 +174,7 @@ const TooltipProvider = () => {
 			return delegate(document.body, eventName, SELECTOR_TRIGGER, () => {
 				dispatch({type: 'hide'});
 
-				restoreTitle(state.target);
+				restoreTitle(state.target as HTMLElement);
```
Comment (bryceosterhaus) [inline modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx:177]: I wasn't quite sure on how to type the reducer's initial state and added this here.

File: modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/reducer.ts [hunk 2]
```diff
@@ -22,7 +22,19 @@ const STATES = {
 
 export {STATES};
 
-export default function reducer(state, action) {
+interface State {
+	current: {show: boolean};
+	nextTarget?: HTMLElement;
+	target?: HTMLElement;
+	timestamp?: number;
```
Comment (bryceosterhaus) [inline modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/reducer.ts:29]: I'm not quite sure the best to to type this. For the argument we pass to the `reducer(` function, the arguments are optional but when we return state from the reducer it will always exist. Is there a more appropriate way to do this? My ts brain is a bit foggy from parental leave

**Comment by bryceosterhaus** [comment]
added a second commit with some updates, not sure if its better this way....

File: modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx [hunk 3]
```diff
@@ -14,6 +14,10 @@
 
 import ClayTooltip from '@clayui/tooltip';
 import {render, useTimeout} from '@liferay/frontend-js-react-web';
+
+// Hack to get around frontend-js-web not being in TS
+// @ts-ignore
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx:19]: Yep, we may end up having to do a lot of this 😢, unless we start at the leaf nodes... (of which `frontend-js-web` is an obvious one).

Because `frontend-js-web` is so "core", we should probably try to port it, at least partially, sooner rather than later. Otherwise all the effectively `any` types that are going to percolate outwards from wherever it is use are going to weaken the guarantees provided by the type system, and probably be painful to clean up too.

File: modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx [hunk 4]
```diff
@@ -34,7 +38,7 @@ const ALIGN_POSITIONS = [
 	'bottom-left',
 	'left',
 	'right',
-];
+] as const;
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx:41]: `as const`: ❤️

File: modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/reducer.ts [hunk 5]
```diff
@@ -22,7 +22,18 @@ const STATES = {
...
+}
+
+type Action =
+	| {type: 'show'; target?: HTMLElement}
+	| {type: 'hide' | 'hideDelayCompleted' | 'showDelayCompleted'};
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/reducer.ts:34]: Probably clearer as:

```ts
type Action =
	| {type: 'hide'}
	| {type: 'hideDelayCompleted'}
	| {type: 'showDelayCompleted'}
	| {type: 'show'; target?: HTMLElement};
```

Given that types are elided from the output, both versions are 0 bytes anyway, so may as well go with the most readable one. `#microoptimization`

Also, sorting, because `#ThisIsLiferay`.

this-is-sparta

Backup gif:

this-is-the-way

File: modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/reducer.ts (hunk 2 above)
Comment (wincent) [reply modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/reducer.ts:29]: If I understand your question correctly, you're saying how we initialize with `{current: STATES.IDLE}` but always set the other properties in our `reducer()` return value, right?

"Always" isn't strictly true, because there is one case we we reset back and:

```ts
return {current: STATES.IDLE};
```

and a couple others where we could conceivably "trick" the reducer into returning an incomplete set of properties, if we dispatched actions in the right order:

```ts
return {
	...state,
	timestamp: Date.now(),
};
```

For me when this kind of situation arises the question isn't "How do I best express this in the type system?" as "Is the pain of typing this a sign that I should be refactoring instead?". In this case, I think it might be worth at least considering that (making sure all the properties are at least present, even if the values are undefined).

Having said that though, I don't actually think what you have here looks too bad, and I really love the way you were able to port 110-line `reducer.js` and 229-line `index.js` to TS with _barely any explicit type annotations at all_. That's awesome. (No doubt some of this is because of the `any` types from `frontend-js-web`, but still...)

File: modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx [hunk 6]
```diff
@@ -158,7 +162,7 @@ const TooltipProvider = () => {
 				document.body,
 				eventName,
 				SELECTOR_TRIGGER,
-				(event) => {
+				(event: MouseEvent & {delegateTarget: HTMLElement}) => {
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx:165]: This is the only sadness in the PR for me, because `delegateTarget` is a non-standard property. But given that we are using `delegate` from `frontend-js-web`, and I doubt that is going to change any time soon, I guess we don't have a choice to continue working like this.

**Review by wincent — COMMENTED** [review]
> Can you give me a sanity check on this just to be sure I'm not doing anything wrong? I'm not super confident on my ts here either, so if you notice anything let me know. I mostly just wanted to test this actually working and it seems to work when deployed for me.

Can't see anything wrong, and given that the changes are almost entirely invisible outside the type system (you literally just added a `|| ''` in one place), it seems pretty safe that if this builds and works when you test it manually, that it is probably good.

File: modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx (hunk 1 above)
Comment (wincent) [reply modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx]: Glad to see you were able to get rid of this cast.

File: modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx (hunk 4 above)
Comment (julien) [reply modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx:41]: off topic comment, but what's the point `const ... as const`?

Comment (julien) [reply modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx:41]: Oh, sorry for asking. 
I guess this is called ["const assertions](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#:~:text=The%20const%20assertion%20allowed%20TypeScript,to%20use%20TypeScript's%20enum%20construct.)?" So now I get the point. Sorry about the noise

Comment (julien) [reply modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx:41]: So is this "similar" to `Object.freeze`?

Comment (wincent) [reply modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx:41]: Yep, const assertions.

"Similar" to `Object.freeze` in the sense that it marks objects/arrays as `readonly` _in the type-system sense_. But _not_ in the runtime sense; the compiler doesn't compile this to `Object.freeze()` or anything under the covers. You can still come along in a (non-TS) JS file and mutate the values at runtime. But at least in TS files, the compiler will error if it sees you trying to write to such a value at compile-time.

The real benefit of `as const`, though, is that it prevents TypeScript from inferring undesirably broad types. For example, given an array like `['foo', 'bar']`, TS will infer a type of `Array<string>`. But with `as const` it will infer type as a read-only _tuple_ (eg. containing exactly 2 elements, immutable, and additionally that the values will be _exactly_ `'foo'` and `'bar'`). So in many places, the narrower/stricter type is exactly what you want and allows TS to detect many potential errors (as well as making reads easier too, requiring less hoop-jumping, casting, and edge-case-handling).

Comment (julien) [reply modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx:41]: Thanks, reading your comment made me understand more things in less time than the article I linked and that I was having a look at.

Comment (wincent) [reply modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.tsx:41]: Russian Novels: They Work™!

**Comment by bryceosterhaus** [comment]
@wincent Thanks for the feedback!

Is it worth making these updates and actually merging these changes? Or do we care much about porting existing js to ts?

**Comment by wincent** [comment]
> Is it worth making these updates and actually merging these changes? Or do we care much about porting existing js to ts?

I would do it, because it fits what I said in Slack last night:

> i think the migration POC you did the other day is a good example of useful work. we should feel free to touch/port anything that infra "owns", and the more we port, the more examples of "best practice" we will have to point at

**Comment by bryceosterhaus** [comment]
ci:test:relevant

**Comment by bryceosterhaus** [comment]
ci:test:sf

**Comment by bryceosterhaus** [comment]
@wincent what is our recommended way to satisfy eslint and SF for the example below. I get a format error saying 
> 22:1  error  imports must be grouped (unexpected blank line before: "frontend-js-web")  @liferay/liferay/group-imports


```ts
import ClayTooltip from '@clayui/tooltip';
import {render, useTimeout} from '@liferay/frontend-js-react-web';

// Hack to get around frontend-js-web not being in TS
// @ts-ignore

import {ALIGN_POSITIONS as POSITIONS, align, delegate} from 'frontend-js-web';
import React, {
	useEffect,
	useLayoutEffect,
	useReducer,
	useRef,
	useState,
} from 'react';
```

**Comment by jbalsas** [comment]
> Is it worth making these updates and actually merging these changes? Or do we care much about porting existing js to ts?

@bryceosterhaus, just echoing @wincent and adding my own 2cts... I would try to progressively migrate our newer APIs to TS in the context of usages here as that should make it also easier for future users and help us further push this change forward.

**Comment by wincent** [comment]
> @wincent what is our recommended way to satisfy eslint and SF for the example below. I get a format error saying
> 
> > 22:1  error  imports must be grouped (unexpected blank line before: "frontend-js-web")  @liferay/liferay/group-imports
> 
> ```ts
> import ClayTooltip from '@clayui/tooltip';
> import {render, useTimeout} from '@liferay/frontend-js-react-web';
> 
> // Hack to get around frontend-js-web not being in TS
> // @ts-ignore
> 
> import {ALIGN_POSITIONS as POSITIONS, align, delegate} from 'frontend-js-web';
> import React, {
> 	useEffect,
> 	useLayoutEffect,
> 	useReducer,
> 	useRef,
> 	useState,
> } from 'react';
> ```

😭 — this looks like an edge case that isn't covered by our rules. They _do_ have logic in them to allow a comment to start a "new group" (see, [we even have a test case for it](https://github.com/liferay/liferay-frontend-projects/blob/11c1b08b8d18bd24eb93c58ee5708bc054d527c4/projects/eslint-config/plugins/liferay/tests/lib/rules/group-imports.js#L191-L201)), so I think you should file a bug in [liferay-frontend-projects](https://github.com/liferay/liferay-frontend-projects). It might be some subtle interaction with another rule (maybe the one that enforces a blank line after comments), but hopefully won't be hard to fix.

The other thing, if you want an immediate workaround, is to "port" `frontend-js-web`'s `index` file to TS (you don't even have to _really_ port it. You can just rename it to `.ts` and mark every as `any`. And then from there we can work on porting bits of it to be properly typed. This is why I said in some other PR or issue or comment somewhere, that we will have to start with the leaf nodes...

**Comment by wincent** [comment]
I went ahead and created [an issue](https://github.com/liferay/liferay-frontend-projects/issues/515) for the formatting problem and am looking into it now.

**Comment by bryceosterhaus** [comment]
@wincent I can't seem to figure out how to mark all the `exports {default as ...` statements as `any`. From what I've read, you can't cast types to export/import statements. I saw examples of using `const someImport: any = require('./some/import');` but that doesn't seem to work with the bundler. Any idea how to gracefully do this?

**Comment by wincent** [comment]
> @wincent I can't seem to figure out how to mark all the `exports {default as ...` statements as `any`. From what I've read, you can't cast types to export/import statements. I saw examples of using `const someImport: any = require('./some/import');` but that doesn't seem to work with the bundler. Any idea how to gracefully do this?

I was initially assuming that doing something like the following (in `frontend-js-web/src/main/resources/META-INF/resources/index.es.js`) would work:

```ts
// ts-nocheck

export {default as AOP} from './liferay/aop/AOP.es';

// etc
```

but maybe you already tried that.

This will be _horribly_ ugly, but might work:

```ts
import {default as _AOP} from './liferay/aop/AOP.es';

export const AOP: any = _AOP;

// etc
```

The thing is, I am kind of surprised that the `import` in that `frontend-js-web` index aren't already coming in as `any` seeing as they're in JS. Maybe that is dependent on [settings](https://www.typescriptlang.org/tsconfig#allowJs) — possibly `"allowJs": true` + `"checkJs": false` would make it work that way... would have to try it out.

Whatever, yet another option: provide `any` types via JSDoc comments in the _JS_ files that are being imported into the index:

```ts
/** @type {any} */
```

(via [this page](https://www.typescriptlang.org/docs/handbook/intro-to-js-ts.html#providing-type-hints-in-js-via-jsdoc)).

Just some ideas given that I haven't pulled down your code so haven't played with it directly yet.

**Comment by wincent** [comment]
@bryceosterhaus, about the formatting problem, fix is with Brian here: https://github.com/brianchandotcom/liferay-portal/pull/101330

**Comment by bryceosterhaus** [comment]
ci:test:sf

**Comment by bryceosterhaus** [comment]
ci:test:relevant

**Comment by bryceosterhaus** [comment]
ci:test:sf

**Comment by bryceosterhaus** [comment]
ci:test:relevant

**Comment by bryceosterhaus** [comment]
ci:forward

**Comment by bryceosterhaus** [comment]
ci:test:relevant
